internal/crd/demand_informer.go

```diff
@@ -0,0 +1,142 @@
+}
+
+// NewLazyDemandInformer constructs a new LazyDemandInformer instance
+func NewLazyDemandInformer(
+	informerFactory ssinformers.SharedInformerFactory,
+	apiExtensionsClient apiextensionsclientset.Interface) *LazyDemandInformer {
+	return &LazyDemandInformer{
+		informerFactory:     informerFactory,
+		apiExtensionsClient: apiExtensionsClient,
+		ready:               make(chan struct{}),
+	}
+}
+
+// Informer returns the informer instance if it is initialized, returns false otherwise
+func (ldi *LazyDemandInformer) Informer() (v1alpha1.DemandInformer, bool) {
+	ldi.lock.RLock()
+	defer ldi.lock.RUnlock()
+	select {
+	case <-ldi.Ready():
+		return ldi.informer, true
+	default:
+		return nil, false
+	}
+}
+
+// Ready returns a channel that will be closed when the informer is initialized
+func (ldi *LazyDemandInformer) Ready() <-chan struct{} {
+	return ldi.ready
+}
+
+// Run starts the goroutine to check for the existence of the demand CRD,
+// and initialize the demand informer if CRD exists
+func (ldi *LazyDemandInformer) Run(ctx context.Context) {
+	if ldi.checkDemandCRDExists(ctx) {
+		return
+	}
+	go func() {
+		_ = wapp.RunWithFatalLogging(ctx, ldi.doStart)
+	}()
+}
+
+func (ldi *LazyDemandInformer) doStart(ctx context.Context) error {
+	t := time.NewTicker(time.Minute)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-t.C:
+			if ldi.checkDemandCRDExists(ctx) {
+				return nil
+			}
+		}
+	}
+}
+
+func (ldi *LazyDemandInformer) checkDemandCRDExists(ctx context.Context) bool {
+	_, ready, err := CheckCRDExists(demandapi.DemandCustomResourceDefinitionName(), ldi.apiExtensionsClient)
+	if err != nil {
+		svc1log.FromContext(ctx).Info("failed to determine if demand CRD exists", svc1log.Stacktrace(err))
+		return false
+	}
+	if ready {
+		svc1log.FromContext(ctx).Info("demand CRD has been initialized. Demand resources can now be created")
+		err = ldi.initializeInformer(ctx)
+		if err != nil {
+			svc1log.FromContext(ctx).Error("failed initializing demand informer", svc1log.Stacktrace(err))
+			return false
+		}
+	}
+	return ready
+}
+
+func (ldi *LazyDemandInformer) initializeInformer(ctx context.Context) error {
+	ldi.lock.Lock()
+	defer ldi.lock.Unlock()
+	informerInterface := ldi.informerFactory.Scaler().V1alpha1().Demands()
+	informer := informerInterface.Informer()
+	ldi.informerFactory.Start(ctx.Done())
+
+	err := retry.Do(ctx, func() error {
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, informerSyncTimeout)
+		defer cancel()
+		if ok := clientcache.WaitForCacheSync(ctxWithTimeout.Done(), informer.HasSynced); !ok {
+			return werror.ErrorWithContextParams(ctx, "timeout syncing informer", werror.SafeParam("timeoutSeconds", informerSyncTimeout.Seconds()))
+		}
+		return nil
+	}, retry.WithMaxAttempts(informerSyncRetryCount), retry.WithInitialBackoff(informerSyncRetryInitialBackoff))
+
+	if err != nil {
+		return err
+	}
+	ldi.informer = informerInterface
+	close(ldi.ready)
```

Comment:
It took me a while to find where this was being done, think you should move it to `doStart` and let this be a pure function that just creates and returns the informer.

Reply:
Then I thought we would need to close in two different places, in `doStart` for the async case, and in `Run`, before spawning the goroutine for the synchronous case